src/common/src/lru.rs

```diff
@@ -0,0 +1,369 @@
+    pub static SEQUENCER: RefCell<Sequencer> = RefCell::new(Sequencer::new(Sequencer::DEFAULT_STEP, Sequencer::DEFAULT_LAG));
+}
+
+static SEQUENCER_DEFAULT_STEP: AtomicSequence = AtomicSequence::new(Sequencer::DEFAULT_STEP);
+static SEQUENCER_DEFAULT_LAG: AtomicSequence = AtomicSequence::new(Sequencer::DEFAULT_LAG);
+
+pub fn init_global_sequencer_args(step: Sequence, lag: Sequence) {
+    SEQUENCER_DEFAULT_STEP.store(step, Ordering::Relaxed);
+    SEQUENCER_DEFAULT_LAG.store(lag, Ordering::Relaxed);
+}
+
+struct LruEntry<K, V>
+where
+    K: Hash + Eq,
+{
+    prev: Option<NonNull<LruEntry<K, V>>>,
+    next: Option<NonNull<LruEntry<K, V>>>,
+    key: MaybeUninit<K>,
+    value: MaybeUninit<V>,
+    hash: u64,
+    sequence: Sequence,
+}
+
+impl<K, V> LruEntry<K, V>
+where
+    K: Hash + Eq,
+{
+    fn key(&self) -> &K {
+        unsafe { self.key.assume_init_ref() }
+    }
+
+    fn value(&self) -> &V {
+        unsafe { self.value.assume_init_ref() }
+    }
+
+    fn value_mut(&mut self) -> &mut V {
+        unsafe { self.value.assume_init_mut() }
+    }
+}
+
+unsafe impl<K, V> Send for LruEntry<K, V> where K: Hash + Eq {}
+unsafe impl<K, V> Sync for LruEntry<K, V> where K: Hash + Eq {}
+
+pub struct LruCache<K, V, S = RandomState, A = Global>
+where
+    K: Hash + Eq,
+    S: BuildHasher + Send + Sync + 'static,
+    A: Clone + Allocator,
+{
+    map: HashTable<NonNull<LruEntry<K, V>>, A>,
+    /// dummy node of the lru linked list
+    dummy: Box<LruEntry<K, V>, A>,
+
+    alloc: A,
+    hash_builder: S,
+}
+
+unsafe impl<K, V, S, A> Send for LruCache<K, V, S, A>
+where
+    K: Hash + Eq,
+    S: BuildHasher + Send + Sync + 'static,
+    A: Clone + Allocator,
+{
+}
+unsafe impl<K, V, S, A> Sync for LruCache<K, V, S, A>
+where
+    K: Hash + Eq,
+    S: BuildHasher + Send + Sync + 'static,
+    A: Clone + Allocator,
+{
+}
+
+impl<K, V> LruCache<K, V>
+where
+    K: Hash + Eq,
+{
+    pub fn unbounded() -> Self {
+        Self::unbounded_with_hasher_in(RandomState::default(), Global)
+    }
+}
+
+impl<K, V, S, A> LruCache<K, V, S, A>
+where
+    K: Hash + Eq,
+    S: BuildHasher + Send + Sync + 'static,
+    A: Clone + Allocator,
+{
+    pub fn unbounded_with_hasher_in(hash_builder: S, alloc: A) -> Self {
+        let map = HashTable::new_in(alloc.clone());
+        let mut dummy = Box::new_in(
+            LruEntry {
+                prev: None,
+                next: None,
+                key: MaybeUninit::uninit(),
+                value: MaybeUninit::uninit(),
+                hash: 0,
+                sequence: Sequence::default(),
+            },
+            alloc.clone(),
+        );
+        let ptr = unsafe { NonNull::new_unchecked(dummy.as_mut() as *mut _) };
+        dummy.next = Some(ptr);
+        dummy.prev = Some(ptr);
+        Self {
+            map,
+            dummy,
+            alloc,
+            hash_builder,
+        }
+    }
+
+    pub fn put(&mut self, key: K, mut value: V) -> Option<V> {
+        unsafe {
```

Comment:
So many `unsafe` in this file 🥵 Please explain the necessities of `unsafe` with some comments in this file.

Reply:
Also, I tend to wrap every linked list operations in `unwrap { ... }` instead of simply wrapping all the function body code. It makes it hard to reason about the safety.

Reply:
The single thread LRU with sequence implementation is basically ported from our original modified lru [repo](https://github.com/risingwavelabs/lru-rs), foyer, and our original block cache implementation (ported from RocksDB). 

A LRU cache is a classic multi-indexer problem, which cannot be achieved easily and cheaply with safe Rust. It requires mutability with shared pointers, O(1) node lookup with address or reference (which cannot be achieved with std linked list) .

Reply:
> It requires mutability with shared pointers

Yeah, got this. But is it possible to reduce the size of `unsafe` block? As mentioned in the 2nd comment.